src/Middleware/RequestLimiter/src/RequestLimiterMiddleware.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.ResourceLimits;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RequestLimiter
+{
+    public class RequestLimiterMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger _logger;
+        private readonly RequestLimiterOptions _options;
+
+        public RequestLimiterMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOptions<RequestLimiterOptions> options)
+        {
+            _next = next;
+            _logger = loggerFactory.CreateLogger<RequestLimiterMiddleware>();
+            _options = options.Value;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            _logger.LogInformation("Resource limiting: " + context.Request.Path);
+
+            var endpoint = context.GetEndpoint();
+            var attributes = endpoint?.Metadata.GetOrderedMetadata<RequestLimitAttribute>();
+
+            if (attributes == null)
+            {
+                await _next.Invoke(context);
+                return;
+            }
+
+            var resourceLeases = new Stack<ResourceLease>();
+            try
+            {
+                foreach (var attribute in attributes)
+                {
+                    if (!string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter != null)
+                    {
+                        throw new InvalidOperationException("Cannot specify both policy and limiter");
+                    }
+
+                    if (string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter == null)
+                    {
+                        if (_options.ResolveDefaultRequestLimit != null)
+                        {
+                            if (!await ApplyLimitAsync(_options.ResolveDefaultRequestLimit(context.RequestServices), context, resourceLeases))
```

Comment:
The default only applies if there's an attribute but no values? There's no way to apply a default to requests without an attribute? AuthZ added a FallbackPolicy that applies to all requests without an attribute.
https://github.com/dotnet/aspnetcore/blob/af96abad0112a90110c0f2d46104fb2e06ed2de2/src/Security/Authorization/Core/src/AuthorizationOptions.cs#L38

Reply:
Great question, I'm not sure which behaviour we want. Or maybe even both?